### ISSUE TITLE
Connection leak on receiving the RateLimitException introduced in 1.9.0

### ISSUE THREAD

**Issue by j-m-x** [auth0/auth0-java#174]
The RateLimitException detection is working perfectly, but unfornately the current implementation forgets to close a response body, hence the following warning in the logs:

`WARNING: A connection to https://<ommited>.auth0.com/ was leaked. Did you forget to close a response body? To see where this was allocated, set the OkHttpClient logger level to FINE: Logger.getLogger(OkHttpClient.class.getName()).setLevel(Level.FINE);`

Simply moving the check on the 429 http status to inside the try-with-resources statement just below is enough to ensure the response body gets properly closed.

**Comment by lbalmaceda** [comment]
`ResponseBody` is a [class that implements](https://square.github.io/okhttp/3.x/okhttp/okhttp3/ResponseBody.html) `Closeable`. Normally, after asking for the body to the response I'd have a `finally` block that closes it, just in case. But this is already done with the new `try` syntax that handles the `Closeable` for you. Examples:

- Success request scenario: https://github.com/auth0/auth0-java/blob/66d189efabfc164b38a35442c11cea7815a79698/src/main/java/com/auth0/net/CustomRequest.java#L64
- Failure request scenario: https://github.com/auth0/auth0-java/blob/66d189efabfc164b38a35442c11cea7815a79698/src/main/java/com/auth0/net/CustomRequest.java#L108


In the event of a `429` status, the `request.body()` is never read, so there's nothing to close. The only methods called up to that point are `request.code()` and `request.headers()`, and none of them are documented saying they should be closed.

**Comment by j-m-x** [comment]
I understand, after reading the docs off the OkHttp lib and agree. I just noticed my logs getting populated by the warning that I mention in the issue I created every time I hit the RateLimitException. Figured I investigate a bit, and after moving the check inside the try catch and creating a local build to run my code, no more warnings. If I understand correctly what you are saying, it means there could be a bug in either the documentation or the implementation of the OkHttp lib this project uses? I haven't gone as far as to check that out yet.

Either way, this change would prevent the warning from showing up. What would you propose?

**Comment by lbalmaceda** [comment]
@j-m-x 
> The caller may read the response body with the response's Response.body method
This only says that the body can be obtained by calling the `response#body` method. Of course if you did that, you need to close the body later.

However on the `ResponseBody` [docs](http://square.github.io/okhttp/3.x/okhttp/okhttp3/ResponseBody.html) I've found this interesting piece:

>  If you invoke Call.execute() or implement Callback.onResponse(okhttp3.Call, okhttp3.Response) you must close this body

So it seems we need to close the response, no matter if we have read the body or not. However, I don't like in your PR that you're doing it after obtaining the response's body which is never used, not even to check if this is a `429` scenario or not. What about moving that to the `createResponseException` method? Since no request body is involved here you might want to do that over the response object instead (which also implements Closeable and would have the same behavior as closing the body). Please, remember to add a test asserting that in the event of a `429` the response is closed. You would do that mocking or spying the response instance and verifying that the close() method was called in the end.

**Comment by lbalmaceda** [comment]
Alternatively, you could move all this into a try/catch/finally block similar to this:


```java
        try {
            if (!response.isSuccessful()) {
                throw createResponseException(response);
            }
            String payload = response.body().string();
            return mapper.readValue(payload, tType);
        } catch (IOException e) {
            throw new APIException("Failed to parse json body", response.code(), e);
        } finally { 
            response.close();
        }
```

**Comment by j-m-x** [comment]
@lbalmaceda 

Thanks for your reply and taking the time to investigate this issue. I agree with you that my current PR is a little crude, I will improve it based on your suggestions, most likely sometime next week.

**Comment by j-m-x** [comment]
Sorry for the long wait, but I've been struggling to decide the best approach here and I would like some feedback on the 3 options I have available as a solution. But before I start I would like to mention the following: For the test I needed to mock the `okhttp3.Response` class in order to do something like this in my test:

```java
Mockito.verify(response, times(1)).close();
```

Unfortunately this class is a final class, so I had to use the `mock-maker-inline` mockito extension ([info](https://github.com/mockito/mockito/wiki/What%27s-new-in-Mockito-2#mock-the-unmockable-opt-in-mocking-of-final-classesmethods)). If you have a different way of mocking or some other solution to get the same result I'll happily change my test.

**Option 1: Fixing this in the `createResponseException` method:**
```java
    protected Auth0Exception createResponseException(Response response) {
        String payload = null;
        try {
            if (response.code() == STATUS_CODE_TOO_MANY_REQUEST) {
                return createRateLimitException(response);
            }

            payload = response.body().string();
            MapType mapType = mapper.getTypeFactory().constructMapType(HashMap.class, String.class, Object.class);
            Map<String, Object> values = mapper.readValue(payload, mapType);
            return new APIException(values, response.code());
        } catch (IOException e) {
            return new APIException(payload, response.code(), e);
        } finally {
            response.close();
        }
    }
```

**Option 2: As you suggested in your last comment, fixing the close in the `parseResponse` method:**
```java
    protected T parseResponse(Response response) throws Auth0Exception {
        try {
            if (!response.isSuccessful()) {
                throw createResponseException(response);
            }

            String payload = response.body().string();
            return mapper.readValue(payload, tType);
        } catch (Auth0Exception e) {
            throw e;
        } catch (IOException e) {
            throw new APIException("Failed to parse json body", response.code(), e);
        } finally {
            response.close();
        }
    }
```
Here I have to explicitly catch the `Auth0Exception` and re-throw it, as this class extends the `IOException`. Without it, every Auth0Exception will be caught as an `IOException` and then re-thrown as an `APIException` with the message "Failed to parse json body"

**Option 3: Personally I would make the method creating the resource that is closeable responsible for closing that resource. In this case the resource is created in `BaseRequest.execute`.**
```java
    public T execute() throws Auth0Exception {
        okhttp3.Request request = createRequest();
        Response response;
        try {
            response = client.newCall(request).execute();
        } catch (IOException e) {
            throw new Auth0Exception("Failed to execute request", e);
        }

        T result = parseResponse(response);
        if (response != null) {
            response.close();
        }
        return result;
    }
```
Optionally I could probably remove all try-with-resources from the `CustomRequest` class with this solution. 

Each of the 3 options pass the new test I added and all existing tests. I would prefer option 3, let me know what you think and I'll update the PR accordingly.

**Comment by lbalmaceda** [comment]
Regarding testing a final class, I'll try to avoid it when possible (I've never tried the approach you suggested though). But note that `ResponseBody` is not final and you can still mock it. And from the [docs](http://square.github.io/okhttp/3.x/okhttp/okhttp3/ResponseBody.html), calling close() on the `Response` instance (final class) is the same as calling it directly to its body: 
> Both this class and Response implement Closeable. Closing a response simply closes its response body. 

If this gets too complicated to test you could extract it to a helper method like [we did here](https://github.com/auth0/Auth0.Android/blob/99eac214759ee4b871dc9ef079a30602740242f5/auth0/src/main/java/com/auth0/android/request/internal/ResponseUtils.java) and test the close() call in there instead. Still since this helper in particular is static, you won't be able to test that the method itself was called. But being the helper method that simple I can make the exception and not require you to test that the request has called it.


Regarding the options you posted, I agree with (3) that the base method should be owning that behavior. But after reading what I wrote above maybe you find yourself stuck on the testing side. Let me know!

**Comment by j-m-x** [comment]
Thanks for your quick reply, glad you agree on the third option. I'll spent some extra time on figuring out a way to test this in a simpler way not mocking a final class. Using the `ResponseBody` is something worth trying.

**Comment by j-m-x** [comment]
I've updated my PR with some minor changes to the third option. I didn't run all tests properly last night and with the code from my last comment it caused some other tests to fail. I have to explicitly catch and re-throw the `Auth0Exception` type in orde to not downgrade all exceptions to an `IOException`. I also realised this is the perfect place for a try-with-resources construction. What I posted last night was not closing the Response on an Exception, which meant my test wasn't any good.

On the testing, mocking a final class is not the nicest thing to do, but with the current setup at least I can guarantee the close is called in every scenario. Trying to check if the body is closed instead of the actual response still requires me the mock the response to throw in my mocked body. Which for me defeats the point of checking the body, as I can just as easily check the response as it is mocked anyway. Not sure if I'm missing the point here. The tests I have now are available in the PR and still use the `mock-maker-inline` setup unfortunately. The tests check if the `BaseRequest.execute()` calls the `response.close()` in a successful context and in three types of errors being `RateLimitException`, `APIException` and `Auth0Exception`.

Let me know what you think and we'll go from there.
